haystack/nodes/preprocessor/merger.py

```diff
@@ -0,0 +1,301 @@
+                               from the merge of the following groups: `[doc1, doc2, doc3]`, `[doc2, doc3, doc4]`, ...
+        :param realign_headlines: whether to update the value of `start_idx` for the document's headlines, if found
+                                  in the metadata. Setting it to False will drop all the headline information found.
+        :param retain_page_number: whether to set the page number to the lowest value in case of mismatch across the
+                                   merged documents. Setting it to False will always drop the page number from the
+                                   merged document.
+        """
+        if not all(doc.content_type == "text" for doc in documents):
+            raise ValueError(
+                "Some of the documents provided are non-textual. Document Merger only works on textual documents."
+            )
+
+        # For safety, as we manipulate the meta
+        documents = deepcopy(documents)
+
+        separator = separator if separator is not None else self.separator
+        window_size = window_size if window_size is not None else self.window_size
+        window_overlap = window_overlap if window_overlap is not None else self.window_overlap
+        realign_headlines = realign_headlines if realign_headlines is not None else self.realign_headlines
+        retain_page_number = retain_page_number if retain_page_number is not None else self.retain_page_number
+        self._validate_window_params(window_size=window_size, window_overlap=window_overlap)
+
+        # Create the groups according to window_size and window_overlap
+        if window_size:
+            groups = [
+                documents[pos : pos + window_size]
+                for pos in range(0, max(1, len(documents) - window_overlap), window_size - window_overlap)
+            ]
+        else:
+            groups = [documents]
+
+        merged_documents = [
+            self.merge(
+                group=group,
+                separator=separator,
+                realign_headlines=realign_headlines,
+                retain_page_number=retain_page_number,
+            )
+            for group in groups
+            if group
+        ]
+        return {"documents": merged_documents}, "output_1"
+
+    def run_batch(  # type: ignore
+        self,
+        documents: List[List[Document]],
+        separator: Optional[str] = None,
+        window_size: Optional[int] = None,
+        window_overlap: Optional[int] = None,
+        realign_headlines: Optional[bool] = None,
+        retain_page_number: Optional[bool] = None,
+    ):
+        """
+        Merges all or a subset of the documents into bigger documents.
+
+        Will retain all metadata that is shared across all documents with the same value
+        (for example, filename might be retained of all documents come from the same file),
+
+        Can treat some special fields separately:
```

Comment:
You can set additional actions for headlines and pages by passing them as parameters: